remotes/docker/resolver.go

```diff
@@ -51,19 +46,36 @@ var (
 	ErrInvalidAuthorization = errors.New("authorization failed")
 )
 
-type dockerResolver struct {
-	credentials func(string) (string, string, error)
-	host        func(string) (string, error)
-	plainHTTP   bool
-	client      *http.Client
-	tracker     StatusTracker
+// Authorizer is used to authorize HTTP requests based on 401 HTTP responses.
+// An Authorizer is responsible for caching tokens or credentials used by
+// requests.
+type Authorizer interface {
+	// Authorize returns the appropriate `Authorization`
+	// header for the given request.
+	//
+	// May return an emptry string or
+	//  "bearer <some bearer token>"
+	//  "basic <base64 encoded credentials>"
+	Authorize(context.Context, *http.Request) (string, error)
```

Comment:
Should this just write the header into the request? There may be authorization schemes that add other headers.

Reply:
I do prefer that, but I was very hesitant to make argument side effecting part of a public interface.  I didn't want to have that argument, but if you also think this case is worth it, I don't mind making that change.

Reply:
Updated, better to just keep the interface less opinionated